lib/mongo/topology.ex

```diff
@@ -0,0 +1,216 @@
+        new_description: new_state.topology
+      })
+      {:reply, :ok, new_state}
+    else
+      {:reply, :ok, new_state}
+    end
+  end
+
+  def handle_cast({:force_check, server_address}, state) do
+    case Map.fetch(state.monitors, server_address) do
+      {:ok, monitor_pid} ->
+        :ok = Monitor.force_check(monitor_pid)
+        {:noreply, state}
+
+      :error ->
+        # ignore force checks on monitors that don't exist
+        {:noreply, state}
+    end
+  end
+
+  defp handle_server_description(state, server_description) do
+    state
+    |> get_and_update_in([:topology],
+                         &TopologyDescription.update(&1, server_description, length(state.seeds)))
+    |> process_events
+    |> reconcile_servers
+  end
+
+  defp process_events({events, state}) do
+    Enum.each(events, fn
+      {:force_check, _} = message ->
+        :ok = GenServer.cast(self, message)
+      {previous, next} ->
+        if previous != next do
+          :ok = Mongo.Events.notify(%ServerDescriptionChangedEvent{
+            address: next.address,
+            topology_pid: self,
+            previous_description: previous,
+            new_description: next
+          })
+        end
+      _ ->
+        :ok
+    end)
+    state
+  end
+
+  defp reconcile_servers(state) do
+    old_addrs = Map.keys(state.monitors)
+    new_addrs = Map.keys(state.topology.servers)
+    added = new_addrs -- old_addrs
+    removed = old_addrs -- new_addrs
+
+    state = Enum.reduce(added, state, fn (address, state) ->
+      server_description = state.topology.servers[address]
+      connopts = connect_opts_from_address(state.opts, address)
+      heartbeat_frequency = state.heartbeat_frequency_ms
+      args = [
+        server_description,
+        self,
+        heartbeat_frequency,
+        Keyword.put(connopts, :pool, DBConnection.Connection)
+      ]
+
+      :ok = Mongo.Events.notify(%ServerOpeningEvent{address: address, topology_pid: self})
+
+      {:ok, pid} = Monitor.start_link(args)
+      {:ok, pool} = DBConnection.start_link(Mongo.Protocol, connopts)
+      %{state | monitors: Map.put(state.monitors, address, pid),
+        connection_pools: Map.put(state.connection_pools, address, pool)}
+    end)
+    Enum.reduce(removed, state, fn (address, state) ->
+      :ok = Mongo.Events.notify(%ServerClosedEvent{address: address, topology_pid: self})
+      :ok = Monitor.stop(state.monitors[address])
+      :ok = GenServer.stop(state.connection_pools[address])
+      %{state | monitors: Map.delete(state.monitors, address),
+        connection_pools: Map.delete(state.connection_pools, address)}
+    end)
+  end
+
+  @connect_opts [:database, :pool]
+  defp connect_opts_from_address(opts, address) do
+    host_opts =
+      "mongodb://" <> address
+      |> URI.parse
+      |> Map.take([:host, :port])
+      |> Enum.into([])
+      |> rename_key(:host, :hostname)
+
+    Keyword.merge(Keyword.take(opts, @connect_opts), host_opts)
```

Comment:
In order to preserver backwards compatibility, this `@connect_opts` attribute will need to be expanded quite a bit. I have already begun doing this for things like ssl and various authentication keys. This will become quite large for no reason, especially as more features are added. I think we should just do a merge without the take and let the different levels take what they need from the keyword list.